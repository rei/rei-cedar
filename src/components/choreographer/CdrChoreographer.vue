<script setup lang="ts">
import { useCssModule, type PropType } from 'vue';
import {
  CdrAbstract,
  CdrKicker,
  CdrTitle,
  CdrBody,
  CdrImg,
  CdrRating,
  CdrLink,
  CdrCard,
  CdrSurface,
  CdrMediaObject,
  CdrLayout,
  CdrPicture,
  CdrButton,
  CdrContainer,
} from '../../lib';
import { choreographerSchema } from '../../types/interfaces';

/** Choreographer is in an experimental stage and should be considered unstable */

defineOptions({
  name: 'CdrChoreographer',
});

const componentMap = {
  abstract: CdrAbstract,
  kicker: CdrKicker,
  title: CdrTitle,
  body: CdrBody,
  image: CdrImg,
  picture: CdrPicture,
  rating: CdrRating,
  link: CdrLink,
  card: CdrCard,
  surface: CdrSurface,
  mediaObject: CdrMediaObject,
  layout: CdrLayout,
  button: CdrButton,
  container: CdrContainer,
};

defineProps({
  schema: { type: Array as PropType<choreographerSchema[]>, default: () => [] },
});

const baseClass = 'cdr-choreographer';
const style = useCssModule();
</script>

<template>
  <component
    v-for="(entry, i) in schema"
    :key="`entry-${i}`"
    :is="entry.type ? componentMap[entry.type as keyof typeof componentMap] : 'div'"
    v-bind="entry?.props"
    :class="style[`${baseClass}__${entry.type}`]"
  >
    <template v-if="Array.isArray(entry?.content)">
      <CdrChoreographer :schema="entry?.content" />
    </template>
    <template v-else>
      {{ entry?.content }}
    </template>
    <template
      v-for="(value, key) in entry.slots"
      #[key]
    >
      <CdrChoreographer
        :key="key"
        :schema="[value]"
      />
    </template>
  </component>
</template>

<style lang="scss" module src="./styles/CdrChoreographer.module.scss"></style>
