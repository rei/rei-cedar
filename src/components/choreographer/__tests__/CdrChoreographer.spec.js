import { shallowMount } from '../../../../test/vue-jest-style-workaround.js';
import CdrChoreographer from '../CdrChoreographer.vue';

describe('CdrChoreographer', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(CdrChoreographer);
    expect(wrapper.element).toMatchSnapshot();
  });

});
