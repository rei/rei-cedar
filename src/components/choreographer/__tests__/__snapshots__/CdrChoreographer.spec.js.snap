// Vitest Snapshot v1

exports[`CdrChoreographer > matches snapshot 1`] = `
<div
  data-v-app=""
>
  
  
</div>
`;
